internal/nix/nixprofile/profile.go

```diff
@@ -88,83 +93,75 @@ func ProfileListItems(
 			return nil, err
 		}
 
-		items[item.unlockedReference] = item
+		items = append(items, item)
 	}
 	return items, nil
 }
 
 type ProfileListIndexArgs struct {
-	// For performance you can reuse the same list in multiple operations if you
+	// For performance, you can reuse the same list in multiple operations if you
 	// are confident index has not changed.
-	List       map[string]*NixProfileListItem
+	Items      []*NixProfileListItem
 	Lockfile   *lock.File
 	Writer     io.Writer
-	Input      *devpkg.Package
+	DevPkg     *devpkg.Package
 	ProfileDir string
 }
 
+// ProfileListIndex returns the index of args.DevPkg in the nix profile specified by args.ProfileDir,
+// or -1 if it's not found. Callers can pass in args.Items to avoid having to call `nix-profile list` again.
 func ProfileListIndex(args *ProfileListIndexArgs) (int, error) {
 	var err error
-	list := args.List
-	if list == nil {
-		list, err = ProfileListItems(args.Writer, args.ProfileDir)
+	items := args.Items
+	if items == nil {
+		items, err = ProfileListItems(args.Writer, args.ProfileDir)
 		if err != nil {
 			return -1, err
 		}
 	}
 
-	inCache, err := args.Input.IsInBinaryCache()
+	inCache, err := args.DevPkg.IsInBinaryCache()
 	if err != nil {
 		return -1, err
 	}
 	if inCache {
-		pathInStore, err := args.Input.Installable()
+		// Packages in cache are added by store path, which means we only need to check
+		// for store path equality to find it.
+		pathInStore, err := args.DevPkg.Installable()
 		if err != nil {
-			return -1, err
+			return -1, errors.Wrapf(err, "failed to get installable for %s", args.DevPkg.String())
 		}
-		for _, item := range list {
+		for _, item := range items {
 			if pathInStore == item.nixStorePath {
 				return item.index, nil
 			}
 		}
+		return -1, errors.Wrap(nix.ErrPackageNotFound, args.DevPkg.String())
```

Comment:
this is a somewhat subtle change that I thought I'd highlight.

If a package is inCache, then it *must* have been added by store path. So it follows that we *must* find a store path match in the items, and if not, then the package is Not Found--that is, there's no need to fall back to the other checks below.

Is this a safe assumption? I think so, because the cache doesn't change much and a package is either in it or not.. right? Please let me know if this isn't as safe an assumption as I think

Reply:
I think that makes sense.

One scenario I was thinking about is (sequentially):
1. Existing devbox user has installed a package.
2. We ship a Devbox update that has Remove Nixpkgs feature enabled.
3. This code here runs. The Package happens to be in the binary cache, BUT it had previously been installed the "old way" in the nix profile.

I think this code should still work, since the store path is input-addressed and would match here.

Reply:
I just tested a few different scenarios and verified this is safe. Most importantly, I tested:

1. User has installed a package in the "old" way. Then the user upgrades devbox and runs a devbox command that triggers the installation path. The package is now going to be installed by store-path. Because the store-paths match, then we do nothing. That is, we keep the "old-style" package installed in the profile, because we found it.  This is OK.

2. Same as above, but the package store path does NOT match. In this case, devbox will end up removing the old package and adding the new package, which is good.